File: Data Structures & Algorithms/max-area-of-island/submission-2.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        rows, cols = len(grid), len(grid[0])
        maxArea = 0

        def bfs(row: int, col: int) -> int:
            queue = deque()
            grid[row][col] = 0
            queue.append((row, col))
            currentMax = 1

            while queue:
                row, col = queue.popleft()
                for dr, dc in directions:
                    nr, nc = row + dr, col + dc
                    if nr < 0 or nc < 0 or nr >= rows or nc >= cols or grid[nr][nc] == 0:
                        continue
                    queue.append((nr, nc))
                    grid[nr][nc] = 0
                    currentMax += 1

            return currentMax
        
        for row in range(rows):
            for col in range(cols):
                if grid[row][col] == 1:
                    maxArea = max(maxArea, bfs(row, col))
        
        return maxArea
        

File: Data Structures & Algorithms/rotting-fruit/submission-2.java
class Pair {
    public int row, col;
    public Pair(int row, int col) {
        this.row = row;
        this.col = col;
    }
}

class Solution {
    
    private static final int[][] directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

    public int orangesRotting(int[][] grid) {
        int rows = grid.length, cols = grid[0].length;
        Deque<Pair> queue = new ArrayDeque<>();

        for(int row = 0; row < rows; row++) {
            for(int col = 0; col < cols; col++) {
                if(grid[row][col] == 2) {queue.offer(new Pair(row, col));}
            }
        }

        int minutes = 0;
        while(!queue.isEmpty()) {
            int levelSize = queue.size();
            for(int i = 0; i < levelSize; i++) {
                Pair popped = queue.removeFirst();
                int row = popped.row, col = popped.col;
                for(int[] dir : directions) {
                    int nr = row + dir[0], nc = col + dir[1];
                    if(nr >= 0 && nr < rows && nc >= 0 && nc < cols && grid[nr][nc] == 1) {
                        queue.offer(new Pair(nr, nc));
                        grid[nr][nc] = 2;
                    }
                }
            }
            if(!queue.isEmpty()) {minutes++;}
        }

        for(int row = 0; row < rows; row++) {
            for(int col = 0; col < cols; col++) {
                if(grid[row][col] == 1) {return -1;}
            }
        }

        return minutes;
    }
}